### PR TITLE
Adding batch_gemm operator

### PR DESCRIPTION
**Pull request by jerryyin**
This PR adds support for batch_gemm operator. batch_gemm implements the BLAS interface: 
`C = alpha*A*B + beta*C`

Note that in the equation, C is both input and output. However, this is not possible in tensroflow python interface. One variable needs to be strictly input or output. The actual python batch_gemm interface is equivalent to `Output = alpha*A*B + beta*C`, in which `beta*C` is viewed as a constant input. In the backward pass, only `alpha*A*B` contributes to the gradient. 
`dA = (dA of gradient of A*B)*alpha`
`dB = (dB of gradient of A*B)*alpha`
`dC does not exist because C is a constant`

-----

Test performed:
//tensorflow:tf_python_api_gen_v1
//tensorflow/tools/pip_package:build_pip_package
//tensorflow/python/kernel_tests:batch_gemm_op_test

### CODE REVIEW

**Comment by whchung** [comment]
@jerryyin as expected the test `//tensorflow/python/kernel_tests:batch_gemm_op_test` fails on CUDA path. to merge the PR, and for future upstreaming purpose, how shall we remedy this?

File: tensorflow/core/kernels/batch_gemm_op.cc [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/kernels/batch_matmul_op_impl.h"
+
+#if GOOGLE_CUDA
+#include "third_party/gpus/cuda/include/cuda.h"
```
Comment (whchung) [inline tensorflow/core/kernels/batch_gemm_op.cc]: @jerryyin do we really need this?

File: tensorflow/core/kernels/batch_gemm_op.cc [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  float alpha_;
+  float beta_;
+};
+
+#define REGISTER_BATCH_GEMM_GPU(TYPE)                                 \
+  REGISTER_KERNEL_BUILDER(                                            \
+      Name("BatchGemm").Device(DEVICE_GPU).TypeConstraint<TYPE>("T"), \
+      BatchGemmOp<GPUDevice, TYPE>);
+
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
```
Comment (whchung) [inline tensorflow/core/kernels/batch_gemm_op.cc]: @jerryyin as the op doesn't really work on CUDA yet, should it only carry `#if TENSORFLOW_USE_ROCM`?

File: tensorflow/python/kernel_tests/batch_gemm_op_test.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    x_t_shape = x_in.shape[:-2] + (x_in.shape[-1], x_in.shape[-2])
+    y_t_shape = y_in.shape[:-2] + (y_in.shape[-1], y_in.shape[-2])
+    x = x_in if not transpose_a else x_in.reshape(x_t_shape)
+    y = y_in if not transpose_b else y_in.reshape(y_t_shape)
+    is_floating = x.dtype != np.int32
+    tol = 100 * np.finfo(x.dtype).eps if is_floating else 0
+    with self.cached_session(use_gpu=is_floating) as sess:
+      # Note: Testing with three dimensions only now
+      z_in = np.ones((x_in.shape[0], x_in.shape[1], y_in.shape[2])).astype(x.dtype)
+      alpha = 1.0
```
Comment (whchung) [inline tensorflow/python/kernel_tests/batch_gemm_op_test.py]: @jerryyin would the test case be too loose? it seems we are still fixing `alpha` be `1.0` and `beta` be `0.0`?

**Comment by jerryyin** [comment]
> @jerryyin as expected the test `//tensorflow/python/kernel_tests:batch_gemm_op_test` fails on CUDA path. to merge the PR, and for future upstreaming purpose, how shall we remedy this?

I need to fix it. My development is done on rocm only, the failure does not sounds surprising to me. Upstream-wise, I don't know if there is a chance for this PR, because `C = alpha*A*B + beta*C` calculation is already achievable with simpler operators. Besides, this seems to not be a user-friendly API, because it requires user to be able to know what dimension exactly is for `C` before the calculation. It can be hard for dynamic dimensions.

**Comment by jerryyin** [comment]
@whchung Both are good points, my thoughts are:

1. We probably need make `Cuda` runnable for now for the sake of Diego's perf comparisons. We don't necessarily have enable it for `Cuda` as the final result.
2. For the forward pass, I can compare more `alpha` and `beta` values. Backward pass is a little more special though. Effectively speaking, backward pass's gradient calculation should only be carried out on `C = alpha*A*B` (there is should not be delta to C, which is both input and output). In this sense, I can compare more `alpha` values but keep `beta` to be 0.

**Comment by deven-amd** [comment]
@jerryyin 

Can you provide an example of how this op will be used?

When I look at the interface for `matmul` I only see inputs for "a" and "b"...wondering why the need to explicity specify the "c" input and how it will be used? 

The TF `matmul` op (https://www.tensorflow.org/api_docs/python/tf/linalg/matmul) already provides "batch matrix multiplication" functionality. Can we re-use that op, instead of introducing a new one? If we can reuse that here, all we would need to change here is the ROCm implementation for that op?

**Comment by whchung** [comment]
@deven-amd you are not in the discussion loop so you might find this PR a bit surprising.

The crux of this PR is to allow non-trivial `alpha/beta` be passed down to rocBLAS so we can reduce small Eigen kernels invoked in certain models we are optimizing.

It appears perhaps due to limitations in cuBLAS, alpha is always 1.0, and beta is 0.0 right now in those TF matmul API. But rocBLAS can do better than that, hence we introduce a new API here. Upstreaming is not a priority at this moment. We just want to validate the functionality and its potential performance improvements.

**Comment by jerryyin** [comment]
@deven-amd Sure, the user API is:
```
c_out = math_ops.batch_gemm(
    x, y, c_in, transpose_a=transpose_a, transpose_b=transpose_b, alpha=alpha, beta=beta)
```
`x, y`: Tensor variables
`c_in`: constant Tensor that have exact same dimension as c_out
`transpose_a, transpose_b`: bool
`alpha, beta`: constant floats

It calculates `Output = alpha*A*B + beta*C`

File: tensorflow/core/kernels/batch_gemm_op.cc (hunk 1 above)
Comment (jerryyin) [reply tensorflow/core/kernels/batch_gemm_op.cc]: Removed it in the latest commit

File: tensorflow/core/kernels/batch_gemm_op.cc (hunk 2 above)
Comment (jerryyin) [reply tensorflow/core/kernels/batch_gemm_op.cc]: Cuda failure is a unit test failure only, and not a problem in kernel. It is related with complex data structure that no kernel is registered for. I get rid of the complex type in the latest commit.

File: tensorflow/python/kernel_tests/batch_gemm_op_test.py (hunk 3 above)
Comment (jerryyin) [reply tensorflow/python/kernel_tests/batch_gemm_op_test.py]: In forward pass, I update both alpha and beta be random floats. In the backward pass, I make alpha random float and beta zero.

**Pull request merged**